src/FastExpressionCompiler/FastExpressionCompiler.cs

```diff
@@ -2172,7 +2176,7 @@ private static bool TryEmitIndexAssign(IndexExpression indexExpr, Type instType,
                 if (!EmitMethodCall(il, method))
                     return false;
 
-                if (parent == ExpressionType.Block && method.ReturnType != typeof(void))
+                if (parent == ExpressionType.Block && closure.CurrentBlock.ResultExpr != expr && method.ReturnType != typeof(void))

```

Comment:
This is a nice use of `BlockInfo`, I had a hope that the current data structure will provide just enough context. Glad that it does.

If we face a limit, we may think about tracking the full stack of expressions, not only Block and separate parent, but it will be more complex to track (push and pop from the stack) and will consume more memory.